In [1]:
import numpy as np
import pandas as pd

# Reading in Data

`data_3_0.txt` is a text file with features seperated by commas. `data_4_2.txt` is a text file with only staging variables, listed in the same order

In [2]:
data3 = pd.read_csv('data_3_0.txt',sep=',', header=0)
#need to convert the diff numbres into, three different categories

/opt/conda/envs/ml/lib/python3.8/site-packages/IPython/core/interactiveshell.py:3145: DtypeWarning: Columns (16,24) have mixed types.Specify dtype option on import or set low_memory=False.
  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,


In [3]:
data3

Site recode ICD-O-3/WHO 2008     Sex Summary stage 2000 (1998+)  \
0                             Breast  Female                    In situ   
1                             Breast  Female                    In situ   
2                             Breast  Female                    In situ   
3                             Breast  Female                    In situ   
4                             Breast  Female                    In situ   
...                              ...     ...                        ...   
1327871                       Breast  Female                  Localized   
1327872                       Breast  Female                  Localized   
1327873                       Breast  Female                   Regional   
1327874                       Breast  Female                   Regional   
1327875                       Breast  Female                  Localized   

                                       Grade  \
0           Poorly differentiated; Grade III   
1        Moderately differentiated; Grade II   
2        Moderately differentiated; Grade II   
3                                    Unknown   
4           Poorly differentiated; Grade III   
...                                      ...   
1327871     Poorly differentiated; Grade III   
1327872         Well differentiated; Grade I   
1327873     Poorly differentiated; Grade III   
1327874     Poorly differentiated; Grade III   
1327875  Moderately differentiated; Grade II   

        Lymph-vascular Invasion (2004+ varying by schema)  \
0                                                Blank(s)   
1                                                Blank(s)   
2                                                Blank(s)   
3                                                Blank(s)   
4                                                Blank(s)   
...                                                   ...   
1327871                                          Blank(s)   
1327872                                          Blank(s)   
1327873                                          Blank(s)   
1327874                                          Blank(s)   
1327875                                          Blank(s)   

        Race recode (W, B, AI, API) Age recode with <1 year olds  \
0                             Black                  65-69 years   
1                             White                  80-84 years   
2                             Black                  55-59 years   
3                             White                  65-69 years   
4                             White                  45-49 years   
...                             ...                          ...   
1327871                       White                  60-64 years   
1327872                       White                  55-59 years   
1327873                       White                  60-64 years   
1327874                       White                  55-59 years   
1327875                       White                  70-74 years   

                       Primary Site - labeled Tumor Size Summary (2016+)  \
0                           C50.9-Breast, NOS                   Blank(s)   
1        C50.4-Upper-outer quadrant of breast                   Blank(s)   
2                           C50.9-Breast, NOS                   Blank(s)   
3                           C50.9-Breast, NOS                   Blank(s)   
4        C50.2-Upper-inner quadrant of breast                   Blank(s)   
...                                       ...                        ...   
1327871                     C50.9-Breast, NOS                   Blank(s)   
1327872  C50.2-Upper-inner quadrant of breast                   Blank(s)   
1327873  C50.2-Upper-inner quadrant of breast                   Blank(s)   
1327874    C50.8-Overlapping lesion of breast                   Blank(s)   
1327875  C50.4-Upper-outer quadrant of breast                   Blank(s)   

        CS tumor size (2004-2015)  ...  \
0                        

In [4]:
list(data3.columns)

['Site recode ICD-O-3/WHO 2008',
 'Sex',
 'Summary stage 2000 (1998+)',
 'Grade',
 'Lymph-vascular Invasion (2004+ varying by schema)',
 'Race recode (W, B, AI, API)',
 'Age recode with <1 year olds',
 'Primary Site - labeled',
 'Tumor Size Summary (2016+)',
 'CS tumor size (2004-2015)',
 'Reason no cancer-directed surgery',
 'Histologic Type ICD-O-3',
 'Behavior recode for analysis',
 'Behavior code ICD-O-3',
 'Regional nodes positive (1988+)',
 'Regional nodes examined (1988+)',
 'Total number of in situ/malignant tumors for patient',
 'CS extension (2004-2015)',
 'Laterality',
 'ER Status Recode Breast Cancer (1990+)',
 'PR Status Recode Breast Cancer (1990+)',
 'COD to site recode',
 'SEER cause-specific death classification',
 'SEER other cause of death classification',
 'Survival months',
 'Survival months flag']

In [5]:
data4 = pd.read_csv('data_4_2.txt',sep=',', header=0)

In [6]:
data4

Breast - Adjusted AJCC 6th Stage (1988-2015)  \
0                                                  0   
1                                                  0   
2                                                  0   
3                                                  0   
4                                                  0   
...                                              ...   
1327871                                            I   
1327872                                            I   
1327873                                          IIB   
1327874                                         IIIC   
1327875                                            I   

        Breast - Adjusted AJCC 6th T (1988-2015)  \
0                                            Tis   
1                                            Tis   
2                                            Tis   
3                                            Tis   
4                                            Tis   
...                                          ...   
1327871                                      T1b   
1327872                                      T1c   
1327873                                       T2   
1327874                                       T2   
1327875                                      T1b   

        Breast - Adjusted AJCC 6th N (1988-2015)  \
0                                             N0   
1                                             N0   
2                                             N0   
3                                             N0   
4                                             N0   
...                                          ...   
1327871                                       N0   
1327872                                       N0   
1327873                                       N1   
1327874                                       N3   
1327875                                       N0   

        Breast - Adjusted AJCC 6th M (1988-2015) Behavior recode for analysis  
0                                             M0                      In situ  
1                                             M0                      In situ  
2                                             M0                      In situ  
3                                             M0                      In situ  
4                                             M0                      In situ  
...                                          ...                          ...  
1327871                                       M0                    Malignant  
1327872                                       M0                    Malignant  
1327873                                       M0                    Malignant  
1327874                                       M0                    Malignant  
1327875                                       M0                    Malignant  

[1327876 rows x 5 columns]

In [7]:
list(data4.columns)

['Breast - Adjusted AJCC 6th Stage (1988-2015)',
 'Breast - Adjusted AJCC 6th T (1988-2015)',
 'Breast - Adjusted AJCC 6th N (1988-2015)',
 'Breast - Adjusted AJCC 6th M (1988-2015)',
 'Behavior recode for analysis']

In [8]:
data4 = data4.drop(columns='Behavior recode for analysis')

# Concat data3 and data4

In [9]:
data3 = pd.concat([data3, data4], axis=1)

In [10]:
list(data3.columns)

['Site recode ICD-O-3/WHO 2008',
 'Sex',
 'Summary stage 2000 (1998+)',
 'Grade',
 'Lymph-vascular Invasion (2004+ varying by schema)',
 'Race recode (W, B, AI, API)',
 'Age recode with <1 year olds',
 'Primary Site - labeled',
 'Tumor Size Summary (2016+)',
 'CS tumor size (2004-2015)',
 'Reason no cancer-directed surgery',
 'Histologic Type ICD-O-3',
 'Behavior recode for analysis',
 'Behavior code ICD-O-3',
 'Regional nodes positive (1988+)',
 'Regional nodes examined (1988+)',
 'Total number of in situ/malignant tumors for patient',
 'CS extension (2004-2015)',
 'Laterality',
 'ER Status Recode Breast Cancer (1990+)',
 'PR Status Recode Breast Cancer (1990+)',
 'COD to site recode',
 'SEER cause-specific death classification',
 'SEER other cause of death classification',
 'Survival months',
 'Survival months flag',
 'Breast - Adjusted AJCC 6th Stage (1988-2015)',
 'Breast - Adjusted AJCC 6th T (1988-2015)',
 'Breast - Adjusted AJCC 6th N (1988-2015)',
 'Breast - Adjusted AJCC 6th M

In [11]:
data3.shape

(1327876, 30)

In [12]:
data3['Sex'].value_counts()

Female    1318896
Male         8980
Name: Sex, dtype: int64

# Variable Specific Preprocessing

In [52]:
for name in data3.columns:
    print('\n'+name)
    print(data3[name].value_counts())


Site recode ICD-O-3/WHO 2008
Breast    1327876
Name: Site recode ICD-O-3/WHO 2008, dtype: int64

Sex
Female    1318896
Male         8980
Name: Sex, dtype: int64

Summary stage 2000 (1998+)
Localized           669337
Regional            318963
In situ             255703
Distant              58158
Unknown/unstaged     25715
Name: Summary stage 2000 (1998+), dtype: int64

Grade
Moderately differentiated; Grade II       506450
Poorly differentiated; Grade III          392618
Well differentiated; Grade I              244723
Unknown                                   156303
Undifferentiated; anaplastic; Grade IV     27782
Name: Grade, dtype: int64

Lymph-vascular Invasion (2004+ varying by schema)
Blank(s)    1327876
Name: Lymph-vascular Invasion (2004+ varying by schema), dtype: int64

Race recode (W, B, AI, API)
White                            1065269
Black                             141670
Asian or Pacific Islander         105440
Unknown                             8589
American Indian/

## Drop unused columns

In [13]:
## cols to drop -- 6 columns
cols = ['Site recode ICD-O-3/WHO 2008',  'Lymph-vascular Invasion (2004+ varying by schema)',  'Tumor Size Summary (2016+)', 'Behavior recode for analysis', 'COD to site recode',
        'SEER cause-specific death classification',  'CS extension (2004-2015)']

In [14]:
data_1 = data3.drop(columns=cols)

In [15]:
data_1.columns

Index(['Sex', 'Summary stage 2000 (1998+)', 'Grade',
       'Race recode (W, B, AI, API)', 'Age recode with <1 year olds',
       'Primary Site - labeled', 'CS tumor size (2004-2015)',
       'Reason no cancer-directed surgery', 'Histologic Type ICD-O-3',
       'Behavior code ICD-O-3', 'Regional nodes positive (1988+)',
       'Regional nodes examined (1988+)',
       'Total number of in situ/malignant tumors for patient', 'Laterality',
       'ER Status Recode Breast Cancer (1990+)',
       'PR Status Recode Breast Cancer (1990+)',
       'SEER other cause of death classification', 'Survival months',
       'Survival months flag', 'Breast - Adjusted AJCC 6th Stage (1988-2015)',
       'Breast - Adjusted AJCC 6th T (1988-2015)',
       'Breast - Adjusted AJCC 6th N (1988-2015)',
       'Breast - Adjusted AJCC 6th M (1988-2015)'],
      dtype='object')

In [120]:
for name in data_1.columns:
    print('\n'+name)
    print(data_1[name].value_counts())


Sex
Female    1318896
Male         8980
Name: Sex, dtype: int64

Summary stage 2000 (1998+)
Localized           669337
Regional            318963
In situ             255703
Distant              58158
Unknown/unstaged     25715
Name: Summary stage 2000 (1998+), dtype: int64

Grade
Moderately differentiated; Grade II       506450
Poorly differentiated; Grade III          392618
Well differentiated; Grade I              244723
Unknown                                   156303
Undifferentiated; anaplastic; Grade IV     27782
Name: Grade, dtype: int64

Race recode (W, B, AI, API)
White                            1065269
Black                             141670
Asian or Pacific Islander         105440
Unknown                             8589
American Indian/Alaska Native       6908
Name: Race recode (W, B, AI, API), dtype: int64

Age recode with <1 year olds
60-64 years    168909
55-59 years    166301
65-69 years    161232
50-54 years    159674
45-49 years    137488
70-74 years    133537
75-

## Removing instances based on features

**NOTE: not going to IMPUTE. Will remove instances that do not have specific features. Otherwise, will give them another category as "Unknown"**

data_4 and data_3 when combined have different outputs initially 1317508 and 1318896 due to the NAs in the additional 4 variables to data_4 that cause it to also drop specific rows

In [19]:
remove = [('Sex', ['Male']), 
          ('Summary stage 2000 (1998+)', ['Unknown/unstaged']), 
          ('Grade', ['Unknown']), 
          ('CS tumor size (2004-2015)', ['Blank(s)', '999', '888', '996', '997', '998', '995', '994', '993', '992', '991', '990', '000']), 
          ('Reason no cancer-directed surgery', ['Unknown; death certificate; or autopsy only (2003+)']),
          ('Regional nodes positive (1988+)', [98, 97, 95]), 
          ('Regional nodes examined (1988+)', [0, 95, 96, 97, 98, 99]),  
          ('Total number of in situ/malignant tumors for patient', ['Unknown']), 
          ('Survival months', ['Unknown']),
          #For some reason I did this separately? 
          ('Survival months flag',['Incomplete dates are available and there cannot be zero days of follow-up',
       'Incomplete dates are available and there could be zero days of follow-up']),
          ('SEER other cause of death classification', ['Dead (attributable to causes other than this cancer dx)', 'Dead (missing/unknown COD)']),
          #Staging Variables -- con catentating them 
          ('Breast - Adjusted AJCC 6th Stage (1988-2015)', ['NA', 'UNK Stage', 'Blank(s)']),
          ('Breast - Adjusted AJCC 6th T (1988-2015)', ['TX', 'T0', 'TX Adjusted', 'Any T, Mets', 'NA', 'Blank(s)']),
          ('Breast - Adjusted AJCC 6th N (1988-2015)', ['NX Adjusted', 'Blank(s)''NA']),
          ('Breast - Adjusted AJCC 6th M (1988-2015)', ['MX', 'NA'])
         ]
#9 features 
count = 0
for variable, value in remove:
    col = data_1[[variable]].replace(value, np.nan)
    data_1[[variable]] = col[[variable]]
    data_2 = data_1.dropna()
    
    count = count + 1
    print(variable+" ")
    print('{}\n'.format(count))
    print(data_2.shape)

#df_scale = df_drop_1[['CS tumor size (2004-2015)']].replace(['Blank(s)'], np.nan)
#df_drop_1[['CS tumor size (2004-2015)']] = df_scale[['CS tumor size (2004-2015)']]
#df_drop_1 = df_drop.dropna()

Sex 
1

(447979, 23)
Summary stage 2000 (1998+) 
2

(447979, 23)
Grade 
3

(447979, 23)
CS tumor size (2004-2015) 
4

(447979, 23)
Reason no cancer-directed surgery 
5

(447979, 23)
Regional nodes positive (1988+) 
6

(447979, 23)
Regional nodes examined (1988+) 
7

(447979, 23)
Total number of in situ/malignant tumors for patient 
8

(447979, 23)
Survival months 
9

(447979, 23)
Survival months flag 
10

(447979, 23)
SEER other cause of death classification 
11

(447979, 23)
Breast - Adjusted AJCC 6th Stage (1988-2015) 
12

(445708, 23)
Breast - Adjusted AJCC 6th T (1988-2015) 
13

(439727, 23)
Breast - Adjusted AJCC 6th N (1988-2015) 
14

(439719, 23)
Breast - Adjusted AJCC 6th M (1988-2015) 
15

(439719, 23)


# Testing

In [19]:
remove = [('Sex', ['Male']), 
          ('Summary stage 2000 (1998+)', ['Unknown/unstaged']), 
          ('Grade', ['Unknown']), 
          ('CS tumor size (2004-2015)', ['Blank(s)', '999', '888', '996', '997', '998', '995', '994', '993', '992', '991', '990', '000']), 
          ('Reason no cancer-directed surgery', ['Unknown; death certificate; or autopsy only (2003+)']),
          ('Regional nodes positive (1988+)', [98, 97, 95]), 
          ('Regional nodes examined (1988+)', [0, 95, 96, 97, 98, 99]),  
          ('Total number of in situ/malignant tumors for patient', ['Unknown']), 
          ('Survival months', ['Unknown']),
          #For some reason I did this separately? 
          ('Survival months flag',['Incomplete dates are available and there cannot be zero days of follow-up',
       'Incomplete dates are available and there could be zero days of follow-up']),
          ('SEER other cause of death classification', ['Dead (attributable to causes other than this cancer dx)', 'Dead (missing/unknown COD)'])
          #Staging Variables -- con catentating them 
          
         ]
#9 features 
count = 0
for variable, value in remove:
    col = data3[[variable]].replace(value, np.nan)
    data3[[variable]] = col[[variable]]
    data0 = data3.dropna()
    
    count = count + 1
    print(variable+" ")
    print('{}\n'.format(count))
    print(data0.shape)

Sex 
1

(1318896, 26)
Summary stage 2000 (1998+) 
2

(1293441, 26)
Grade 
3

(1155570, 26)
CS tumor size (2004-2015) 
4

(727211, 26)
Reason no cancer-directed surgery 
5

(726825, 26)
Regional nodes positive (1988+) 
6

(589208, 26)
Regional nodes examined (1988+) 
7

(584130, 26)
Total number of in situ/malignant tumors for patient 
8

(584127, 26)
Survival months 
9

(584127, 26)
Survival months flag 
10

(497097, 26)
SEER other cause of death classification 
11

(448215, 26)


# End Testing

In [20]:
data_2.columns

Index(['Sex', 'Summary stage 2000 (1998+)', 'Grade',
       'Race recode (W, B, AI, API)', 'Age recode with <1 year olds',
       'Primary Site - labeled', 'CS tumor size (2004-2015)',
       'Reason no cancer-directed surgery', 'Histologic Type ICD-O-3',
       'Behavior code ICD-O-3', 'Regional nodes positive (1988+)',
       'Regional nodes examined (1988+)',
       'Total number of in situ/malignant tumors for patient', 'Laterality',
       'ER Status Recode Breast Cancer (1990+)',
       'PR Status Recode Breast Cancer (1990+)',
       'SEER other cause of death classification', 'Survival months',
       'Survival months flag', 'Breast - Adjusted AJCC 6th Stage (1988-2015)',
       'Breast - Adjusted AJCC 6th T (1988-2015)',
       'Breast - Adjusted AJCC 6th N (1988-2015)',
       'Breast - Adjusted AJCC 6th M (1988-2015)'],
      dtype='object')

In [21]:
data_2.shape

(439719, 23)

data-2.txt shape is 448215 rows
data-3.txt shape is 439719 rows

In [22]:
for name in data_2.columns:
    print('\n'+name)
    print(data_2[name].value_counts())


Sex
Female    439719
Name: Sex, dtype: int64

Summary stage 2000 (1998+)
Localized    274633
Regional     132744
In situ       31132
Distant        1210
Name: Summary stage 2000 (1998+), dtype: int64

Grade
Moderately differentiated; Grade II       190719
Poorly differentiated; Grade III          146821
Well differentiated; Grade I               95908
Undifferentiated; anaplastic; Grade IV      6271
Name: Grade, dtype: int64

Race recode (W, B, AI, API)
White                            349122
Black                             43230
Asian or Pacific Islander         43027
American Indian/Alaska Native      2266
Unknown                            2074
Name: Race recode (W, B, AI, API), dtype: int64

Age recode with <1 year olds
60-64 years    62172
55-59 years    60786
50-54 years    59068
65-69 years    55998
45-49 years    52318
70-74 years    41182
40-44 years    33807
75-79 years    28490
80-84 years    15437
35-39 years    15009
85+ years       6869
30-34 years     6274
25-29 years

## Dropping the cause of death variable and survival months

In [23]:
data_3 = data_2.drop(columns=['SEER other cause of death classification', 'Survival months flag'])

In [24]:
data_3.columns

Index(['Sex', 'Summary stage 2000 (1998+)', 'Grade',
       'Race recode (W, B, AI, API)', 'Age recode with <1 year olds',
       'Primary Site - labeled', 'CS tumor size (2004-2015)',
       'Reason no cancer-directed surgery', 'Histologic Type ICD-O-3',
       'Behavior code ICD-O-3', 'Regional nodes positive (1988+)',
       'Regional nodes examined (1988+)',
       'Total number of in situ/malignant tumors for patient', 'Laterality',
       'ER Status Recode Breast Cancer (1990+)',
       'PR Status Recode Breast Cancer (1990+)', 'Survival months',
       'Breast - Adjusted AJCC 6th Stage (1988-2015)',
       'Breast - Adjusted AJCC 6th T (1988-2015)',
       'Breast - Adjusted AJCC 6th N (1988-2015)',
       'Breast - Adjusted AJCC 6th M (1988-2015)'],
      dtype='object')

## Changing Histologic Type ICD-O-3 variable

In [25]:
import math

In [26]:
data_3['histo'] = data_3['Histologic Type ICD-O-3'].apply(lambda x: math.floor(x/10))

## Replacing TMN Staging Variables with Numbers

In [27]:
data_3['Breast - Adjusted AJCC 6th T (1988-2015)'] = data_3['Breast - Adjusted AJCC 6th T (1988-2015)'].replace(['T0', 'Tis', 'T1mic', 'T1a', 'T1b', 'T1c', 'T2', 'T3', 'T4a', 'T4b', 'T4c', 'T4d'], 
                                                           [0, 0, 0.1, 0.25, 0.75, 1.5, 3.0, 5.0, 6.0, 7.0, 8.0, 9.0])

In [28]:
data_3['Breast - Adjusted AJCC 6th N (1988-2015)'] = data_3['Breast - Adjusted AJCC 6th N (1988-2015)'].replace(['N0', 'N1', 'N2', 'N3'],[0, 2, 6.5, 10])

In [29]:
data_3['Breast - Adjusted AJCC 6th M (1988-2015)'] = data_3['Breast - Adjusted AJCC 6th M (1988-2015)'].replace('M0', 0)

## Processing to Binary Label

Survivability will depend on whether the patient lived longer than 5 years (aka 60 months) after first diagnosis as based off previous papers doing this same task). 

In [30]:
target = data_3[['Survival months']]

In [31]:
target

Survival months
5                  0123
10                 0085
33                 0061
47                 0162
50                 0143
...                 ...
1327866            0156
1327867            0149
1327869            0126
1327872            0149
1327874            0154

[439719 rows x 1 columns]

In [32]:
target_1 = pd.to_numeric(target['Survival months']).to_frame()

In [33]:
target_1

Survival months
5                    123
10                    85
33                    61
47                   162
50                   143
...                  ...
1327866              156
1327867              149
1327869              126
1327872              149
1327874              154

[439719 rows x 1 columns]

In [34]:
target_1[target_1['Survival months'] < 60] = -1 #did not survive

In [35]:
target_1[target_1['Survival months'] > 60] = 1 #survived

In [36]:
target_1[target_1['Survival months'] == 60] = 1 #survived

In [37]:
target_1

Survival months
5                      1
10                     1
33                     1
47                     1
50                     1
...                  ...
1327866                1
1327867                1
1327869                1
1327872                1
1327874                1

[439719 rows x 1 columns]

In [38]:
target_1['Survival months'].value_counts()

 1    278714
-1    161005
Name: Survival months, dtype: int64

In [39]:
281919+166296

448215

1 = survived (lived 5 years or longer) --> 281919 instances

0 = did not survive (lived less than 5 years) --> 166296 instances

In [40]:
data_4 = data_3.drop(columns=['Survival months', 'Sex'])

In [41]:
data_4

Summary stage 2000 (1998+)                                   Grade  \
5                          In situ  Undifferentiated; anaplastic; Grade IV   
10                         In situ  Undifferentiated; anaplastic; Grade IV   
33                         In situ        Poorly differentiated; Grade III   
47                         In situ  Undifferentiated; anaplastic; Grade IV   
50                         In situ     Moderately differentiated; Grade II   
...                            ...                                     ...   
1327866                   Regional     Moderately differentiated; Grade II   
1327867                  Localized     Moderately differentiated; Grade II   
1327869                  Localized     Moderately differentiated; Grade II   
1327872                  Localized            Well differentiated; Grade I   
1327874                   Regional        Poorly differentiated; Grade III   

        Race recode (W, B, AI, API) Age recode with <1 year olds  \
5                             White                    85+ years   
10                            White                  60-64 years   
33        Asian or Pacific Islander                  55-59 years   
47                            White                  75-79 years   
50        Asian or Pacific Islander                  60-64 years   
...                             ...                          ...   
1327866                       White                  50-54 years   
1327867                       White                  65-69 years   
1327869                       White                  30-34 years   
1327872                       White                  55-59 years   
1327874                       White                  55-59 years   

                       Primary Site - labeled CS tumor size (2004-2015)  \
5                           C50.9-Breast, NOS                       018   
10            C50.1-Central portion of breast                       025   
33       C50.4-Upper-outer quadrant of breast                       030   
47       C50.4-Upper-outer quadrant of breast                       005   
50       C50.2-Upper-inner quadrant of breast                       033   
...                                       ...                       ...   
1327866  C50.4-Upper-outer quadrant of breast                       012   
1327867    C50.8-Overlapping lesion of breast                       012   
1327869  C50.2-Upper-inner quadrant of breast                       015   
1327872  C50.2-Upper-inner quadrant of breast                       020   
1327874    C50.8-Overlapping lesion of breast                       035   

        Reason no cancer-directed surgery  Histologic Type ICD-O-3  \
5                       Surgery performed                      850   
10                      Surgery performed                      852   
33                      Surgery performed                      852   
47                      Surgery performed                      850   
50                      Surgery performed                      852   
...                                   ...                      ...   
1327866                 Surgery performed                      852   
1327867                 Surgery performed                      850   
1327869                 Surgery performed                      850   
1327872                 Surgery performed                      852   
1327874                 Surgery performed                      850   

        Behavior code ICD-O-3  Regional nodes positive (1988+)  \
5                     In situ                              0.0   
10                    In situ                              0.0   
33                    In situ                              0.0   
47                    In situ                              0.0   
50                    In situ                              0.0   
...                       ...                              ...   
1327866             Malignant                     

In [42]:
data_4.columns

Index(['Summary stage 2000 (1998+)', 'Grade', 'Race recode (W, B, AI, API)',
       'Age recode with <1 year olds', 'Primary Site - labeled',
       'CS tumor size (2004-2015)', 'Reason no cancer-directed surgery',
       'Histologic Type ICD-O-3', 'Behavior code ICD-O-3',
       'Regional nodes positive (1988+)', 'Regional nodes examined (1988+)',
       'Total number of in situ/malignant tumors for patient', 'Laterality',
       'ER Status Recode Breast Cancer (1990+)',
       'PR Status Recode Breast Cancer (1990+)',
       'Breast - Adjusted AJCC 6th Stage (1988-2015)',
       'Breast - Adjusted AJCC 6th T (1988-2015)',
       'Breast - Adjusted AJCC 6th N (1988-2015)',
       'Breast - Adjusted AJCC 6th M (1988-2015)'],
      dtype='object')

# Splitting into Train/Test 
Because you can't have data leak -- where information about the testing set is leaked into the training set? they have to be scaled independently

## Changing y-variable

In [43]:
target = pd.DataFrame(target_1.to_numpy(), columns=['Survival months'])
target

Survival months
0                     1
1                     1
2                     1
3                     1
4                     1
...                 ...
439714                1
439715                1
439716                1
439717                1
439718                1

[439719 rows x 1 columns]

In [49]:
data_5 = pd.DataFrame(data_4.to_numpy(), columns=['Summary stage 2000 (1998+)', 'Grade', 'Race recode (W, B, AI, API)',
       'Age recode with <1 year olds', 'Primary Site - labeled',
       'CS tumor size (2004-2015)', 'Reason no cancer-directed surgery',
       'Histologic Type ICD-O-3', 'Behavior code ICD-O-3',
       'Regional nodes positive (1988+)', 'Regional nodes examined (1988+)',
       'Total number of in situ/malignant tumors for patient', 'Laterality',
       'ER Status Recode Breast Cancer (1990+)',
       'PR Status Recode Breast Cancer (1990+)',
       'Breast - Adjusted AJCC 6th Stage (1988-2015)',
       'Breast - Adjusted AJCC 6th T (1988-2015)',
       'Breast - Adjusted AJCC 6th N (1988-2015)',
       'Breast - Adjusted AJCC 6th M (1988-2015)'])
data_6 = pd.concat([data_5, target], axis=1)
data_6

Summary stage 2000 (1998+)                                   Grade  \
0                         In situ  Undifferentiated; anaplastic; Grade IV   
1                         In situ  Undifferentiated; anaplastic; Grade IV   
2                         In situ        Poorly differentiated; Grade III   
3                         In situ  Undifferentiated; anaplastic; Grade IV   
4                         In situ     Moderately differentiated; Grade II   
...                           ...                                     ...   
439714                   Regional     Moderately differentiated; Grade II   
439715                  Localized     Moderately differentiated; Grade II   
439716                  Localized     Moderately differentiated; Grade II   
439717                  Localized            Well differentiated; Grade I   
439718                   Regional        Poorly differentiated; Grade III   

       Race recode (W, B, AI, API) Age recode with <1 year olds  \
0                            White                    85+ years   
1                            White                  60-64 years   
2        Asian or Pacific Islander                  55-59 years   
3                            White                  75-79 years   
4        Asian or Pacific Islander                  60-64 years   
...                            ...                          ...   
439714                       White                  50-54 years   
439715                       White                  65-69 years   
439716                       White                  30-34 years   
439717                       White                  55-59 years   
439718                       White                  55-59 years   

                      Primary Site - labeled CS tumor size (2004-2015)  \
0                          C50.9-Breast, NOS                       018   
1            C50.1-Central portion of breast                       025   
2       C50.4-Upper-outer quadrant of breast                       030   
3       C50.4-Upper-outer quadrant of breast                       005   
4       C50.2-Upper-inner quadrant of breast                       033   
...                                      ...                       ...   
439714  C50.4-Upper-outer quadrant of breast                       012   
439715    C50.8-Overlapping lesion of breast                       012   
439716  C50.2-Upper-inner quadrant of breast                       015   
439717  C50.2-Upper-inner quadrant of breast                       020   
439718    C50.8-Overlapping lesion of breast                       035   

       Reason no cancer-directed surgery Histologic Type ICD-O-3  \
0                      Surgery performed                     850   
1                      Surgery performed                     852   
2                      Surgery performed                     852   
3                      Surgery performed                     850   
4                      Surgery performed                     852   
...                                  ...                     ...   
439714                 Surgery performed                     852   
439715                 Surgery performed                     850   
439716                 Surgery performed                     850   
439717                 Surgery performed                     852   
439718                 Surgery performed                     850   

       Behavior code ICD-O-3 Regional nodes positive (1988+)  \
0                    In situ                               0   
1                    In situ                               0   
2                    In situ                               0   
3                    In situ                               0   
4                    In situ                               0   
...                      ...                             ...   
439714             Malignant                              10   
439715             Malignant                               

## Resampling
having the same number of survived and did not survive

In [50]:
data_6['Survival months'].value_counts()

 1    278714
-1    161005
Name: Survival months, dtype: int64

In [51]:
low = data_6[['Survival months']].value_counts().min()
low

161005

Choosing the number of random samples that have (data-2 166296)  (data-3 161005) (data-1 171866) for the survived (1) cases

In [52]:
samples = data_6[data_6['Survival months'] == 1].index.values
index_range = range(samples.shape[0])
indexes = np.random.choice(index_range, size=low, replace=False)
subsample =  samples[indexes].tolist()
d_1 = data_6.loc[subsample, :]

In [53]:
d_1

Summary stage 2000 (1998+)                                Grade  \
99395                   Localized     Poorly differentiated; Grade III   
438302                  Localized  Moderately differentiated; Grade II   
14227                     In situ  Moderately differentiated; Grade II   
201514                  Localized         Well differentiated; Grade I   
403716                  Localized         Well differentiated; Grade I   
...                           ...                                  ...   
343737                  Localized  Moderately differentiated; Grade II   
334564                  Localized  Moderately differentiated; Grade II   
259010                  Localized         Well differentiated; Grade I   
263628                  Localized         Well differentiated; Grade I   
184039                   Regional     Poorly differentiated; Grade III   

       Race recode (W, B, AI, API) Age recode with <1 year olds  \
99395    Asian or Pacific Islander                  60-64 years   
438302                       White                  50-54 years   
14227    Asian or Pacific Islander                  50-54 years   
201514                       White                  60-64 years   
403716                       White                  55-59 years   
...                            ...                          ...   
343737                       Black                  60-64 years   
334564                       White                  60-64 years   
259010                       White                  80-84 years   
263628                       White                  50-54 years   
184039                       White                  65-69 years   

                      Primary Site - labeled CS tumor size (2004-2015)  \
99395   C50.4-Upper-outer quadrant of breast                       028   
438302  C50.4-Upper-outer quadrant of breast                       008   
14227   C50.5-Lower-outer quadrant of breast                       025   
201514  C50.4-Upper-outer quadrant of breast                       024   
403716  C50.3-Lower-inner quadrant of breast                       016   
...                                      ...                       ...   
343737  C50.4-Upper-outer quadrant of breast                       013   
334564  C50.3-Lower-inner quadrant of breast                       021   
259010                     C50.9-Breast, NOS                       009   
263628  C50.4-Upper-outer quadrant of breast                       007   
184039    C50.8-Overlapping lesion of breast                       017   

       Reason no cancer-directed surgery Histologic Type ICD-O-3  \
99395                  Surgery performed                     850   
438302                 Surgery performed                     850   
14227                  Surgery performed                     852   
201514                 Surgery performed                     852   
403716                 Surgery performed                     850   
...                                  ...                     ...   
343737                 Surgery performed                     850   
334564                 Surgery performed                     852   
259010                 Surgery performed                     850   
263628                 Surgery performed                     852   
184039                 Surgery performed                     850   

       Behavior code ICD-O-3 Regional nodes positive (1988+)  \
99395              Malignant                               0   
438302             Malignant                               0   
14227                In situ                               0   
201514             Malignant                               0   
403716             Malignant                               0   
...                      ...                             ...   
343737             Malignant                               0   
334564             Malignant                               0   
259010             Malignant   

In [54]:
samples_0 = data_6[data_6['Survival months'] == -1].index.values
d_0 = data_6.loc[samples_0, :]

In [55]:
d_0

Summary stage 2000 (1998+)                                Grade  \
6                         In situ  Moderately differentiated; Grade II   
17                        In situ     Poorly differentiated; Grade III   
18                        In situ     Poorly differentiated; Grade III   
24                        In situ  Moderately differentiated; Grade II   
33                        In situ  Moderately differentiated; Grade II   
...                           ...                                  ...   
439675                  Localized         Well differentiated; Grade I   
439679                  Localized     Poorly differentiated; Grade III   
439684                   Regional  Moderately differentiated; Grade II   
439695                  Localized  Moderately differentiated; Grade II   
439705                  Localized  Moderately differentiated; Grade II   

       Race recode (W, B, AI, API) Age recode with <1 year olds  \
6        Asian or Pacific Islander                  60-64 years   
17                           Black                  50-54 years   
18       Asian or Pacific Islander                  55-59 years   
24                           Black                  55-59 years   
33                           White                  40-44 years   
...                            ...                          ...   
439675                       White                  65-69 years   
439679                       White                  80-84 years   
439684   Asian or Pacific Islander                  50-54 years   
439695   Asian or Pacific Islander                  70-74 years   
439705                       White                  45-49 years   

                      Primary Site - labeled CS tumor size (2004-2015)  \
6       C50.4-Upper-outer quadrant of breast                       008   
17      C50.4-Upper-outer quadrant of breast                       025   
18        C50.8-Overlapping lesion of breast                       020   
24                         C50.9-Breast, NOS                       050   
33      C50.2-Upper-inner quadrant of breast                       018   
...                                      ...                       ...   
439675  C50.4-Upper-outer quadrant of breast                       008   
439679  C50.4-Upper-outer quadrant of breast                       017   
439684       C50.1-Central portion of breast                       040   
439695    C50.8-Overlapping lesion of breast                       008   
439705                     C50.9-Breast, NOS                       015   

       Reason no cancer-directed surgery Histologic Type ICD-O-3  \
6                      Surgery performed                     852   
17                     Surgery performed                     852   
18                     Surgery performed                     850   
24                     Surgery performed                     852   
33                     Surgery performed                     850   
...                                  ...                     ...   
439675                 Surgery performed                     850   
439679                 Surgery performed                     850   
439684                 Surgery performed                     850   
439695                 Surgery performed                     850   
439705                 Surgery performed                     850   

       Behavior code ICD-O-3 Regional nodes positive (1988+)  \
6                    In situ                               0   
17                   In situ                               0   
18                   In situ                               0   
24                   In situ                               0   
33                   In situ                               0   
...                      ...                             ...   
439675             Malignant                               0   
439679             Malignant                               0   
439684             Malignant   

In [56]:
d_final = pd.concat([d_1, d_0])

In [57]:
d_final[['Survival months']].value_counts()

Survival months
 1                 161005
-1                 161005
dtype: int64

In [58]:
d_final

Summary stage 2000 (1998+)                                Grade  \
99395                   Localized     Poorly differentiated; Grade III   
438302                  Localized  Moderately differentiated; Grade II   
14227                     In situ  Moderately differentiated; Grade II   
201514                  Localized         Well differentiated; Grade I   
403716                  Localized         Well differentiated; Grade I   
...                           ...                                  ...   
439675                  Localized         Well differentiated; Grade I   
439679                  Localized     Poorly differentiated; Grade III   
439684                   Regional  Moderately differentiated; Grade II   
439695                  Localized  Moderately differentiated; Grade II   
439705                  Localized  Moderately differentiated; Grade II   

       Race recode (W, B, AI, API) Age recode with <1 year olds  \
99395    Asian or Pacific Islander                  60-64 years   
438302                       White                  50-54 years   
14227    Asian or Pacific Islander                  50-54 years   
201514                       White                  60-64 years   
403716                       White                  55-59 years   
...                            ...                          ...   
439675                       White                  65-69 years   
439679                       White                  80-84 years   
439684   Asian or Pacific Islander                  50-54 years   
439695   Asian or Pacific Islander                  70-74 years   
439705                       White                  45-49 years   

                      Primary Site - labeled CS tumor size (2004-2015)  \
99395   C50.4-Upper-outer quadrant of breast                       028   
438302  C50.4-Upper-outer quadrant of breast                       008   
14227   C50.5-Lower-outer quadrant of breast                       025   
201514  C50.4-Upper-outer quadrant of breast                       024   
403716  C50.3-Lower-inner quadrant of breast                       016   
...                                      ...                       ...   
439675  C50.4-Upper-outer quadrant of breast                       008   
439679  C50.4-Upper-outer quadrant of breast                       017   
439684       C50.1-Central portion of breast                       040   
439695    C50.8-Overlapping lesion of breast                       008   
439705                     C50.9-Breast, NOS                       015   

       Reason no cancer-directed surgery Histologic Type ICD-O-3  \
99395                  Surgery performed                     850   
438302                 Surgery performed                     850   
14227                  Surgery performed                     852   
201514                 Surgery performed                     852   
403716                 Surgery performed                     850   
...                                  ...                     ...   
439675                 Surgery performed                     850   
439679                 Surgery performed                     850   
439684                 Surgery performed                     850   
439695                 Surgery performed                     850   
439705                 Surgery performed                     850   

       Behavior code ICD-O-3 Regional nodes positive (1988+)  \
99395              Malignant                               0   
438302             Malignant                               0   
14227                In situ                               0   
201514             Malignant                               0   
403716             Malignant                               0   
...                      ...                             ...   
439675             Malignant                               0   
439679             Malignant                               0   
439684             Malignant   

In [59]:
d_shuffle = d_final.sample(frac=1)

In [60]:
d_shuffle

Summary stage 2000 (1998+)                                Grade  \
376349                   Regional         Well differentiated; Grade I   
80511                    Regional     Poorly differentiated; Grade III   
17513                     In situ     Poorly differentiated; Grade III   
272156                   Regional     Poorly differentiated; Grade III   
193740                   Regional     Poorly differentiated; Grade III   
...                           ...                                  ...   
405886                    Distant  Moderately differentiated; Grade II   
33605                    Regional  Moderately differentiated; Grade II   
414514                  Localized         Well differentiated; Grade I   
327832                  Localized     Poorly differentiated; Grade III   
41036                   Localized  Moderately differentiated; Grade II   

       Race recode (W, B, AI, API) Age recode with <1 year olds  \
376349                       White                  75-79 years   
80511                        White                  75-79 years   
17513                        White                  45-49 years   
272156                       Black                  50-54 years   
193740                       White                  40-44 years   
...                            ...                          ...   
405886                       White                  40-44 years   
33605                        White                    85+ years   
414514                       White                  75-79 years   
327832   Asian or Pacific Islander                  45-49 years   
41036                        White                  40-44 years   

                      Primary Site - labeled CS tumor size (2004-2015)  \
376349  C50.5-Lower-outer quadrant of breast                       016   
80511     C50.8-Overlapping lesion of breast                       055   
17513   C50.2-Upper-inner quadrant of breast                       032   
272156  C50.4-Upper-outer quadrant of breast                       024   
193740                     C50.9-Breast, NOS                       026   
...                                      ...                       ...   
405886       C50.1-Central portion of breast                       084   
33605   C50.4-Upper-outer quadrant of breast                       014   
414514  C50.5-Lower-outer quadrant of breast                       016   
327832  C50.4-Upper-outer quadrant of breast                       015   
41036   C50.4-Upper-outer quadrant of breast                       040   

       Reason no cancer-directed surgery Histologic Type ICD-O-3  \
376349                 Surgery performed                     850   
80511                  Surgery performed                     850   
17513                  Surgery performed                     850   
272156                 Surgery performed                     850   
193740                 Surgery performed                     850   
...                                  ...                     ...   
405886                 Surgery performed                     852   
33605                  Surgery performed                     852   
414514                 Surgery performed                     850   
327832                 Surgery performed                     850   
41036                  Surgery performed                     850   

       Behavior code ICD-O-3 Regional nodes positive (1988+)  \
376349             Malignant                               1   
80511              Malignant                               1   
17513                In situ                               0   
272156             Malignant                              31   
193740             Malignant                               1   
...                      ...                             ...   
405886             Malignant                              18   
33605              Malignant                               1   
414514             Malignant   

## Split

In [62]:
np.random.seed(42)
from sklearn.model_selection import train_test_split
train, test = train_test_split(d_shuffle, test_size=0.2, random_state=42)
#y = d_shuffle[['Survival months']]
#x = d_shuffle.drop(columns='Survival months')
#print(y.shape, x.shape)

In [63]:
print(train.shape, test.shape)

(257608, 20) (64402, 20)


In [64]:
test

Summary stage 2000 (1998+)                                Grade  \
300957                  Localized  Moderately differentiated; Grade II   
58456                   Localized     Poorly differentiated; Grade III   
429774                   Regional  Moderately differentiated; Grade II   
226774                  Localized  Moderately differentiated; Grade II   
194850                   Regional  Moderately differentiated; Grade II   
...                           ...                                  ...   
204005                   Regional     Poorly differentiated; Grade III   
77221                   Localized     Poorly differentiated; Grade III   
48914                   Localized         Well differentiated; Grade I   
318688                  Localized         Well differentiated; Grade I   
73757                   Localized     Poorly differentiated; Grade III   

       Race recode (W, B, AI, API) Age recode with <1 year olds  \
300957                       White                  60-64 years   
58456                        White                  60-64 years   
429774   Asian or Pacific Islander                  40-44 years   
226774                       White                  70-74 years   
194850                       White                  45-49 years   
...                            ...                          ...   
204005   Asian or Pacific Islander                  45-49 years   
77221                        White                  65-69 years   
48914                        White                  80-84 years   
318688                       White                  45-49 years   
73757                        White                  60-64 years   

                      Primary Site - labeled CS tumor size (2004-2015)  \
300957    C50.8-Overlapping lesion of breast                       035   
58456                      C50.9-Breast, NOS                       011   
429774       C50.1-Central portion of breast                       030   
226774    C50.8-Overlapping lesion of breast                       009   
194850  C50.4-Upper-outer quadrant of breast                       025   
...                                      ...                       ...   
204005  C50.4-Upper-outer quadrant of breast                       045   
77221   C50.4-Upper-outer quadrant of breast                       032   
48914                      C50.9-Breast, NOS                       008   
318688  C50.4-Upper-outer quadrant of breast                       004   
73757   C50.4-Upper-outer quadrant of breast                       029   

       Reason no cancer-directed surgery Histologic Type ICD-O-3  \
300957                 Surgery performed                     902   
58456                  Surgery performed                     850   
429774                 Surgery performed                     850   
226774                 Surgery performed                     850   
194850                 Surgery performed                     852   
...                                  ...                     ...   
204005                 Surgery performed                     852   
77221                  Surgery performed                     850   
48914                  Surgery performed                     850   
318688                 Surgery performed                     821   
73757                  Surgery performed                     850   

       Behavior code ICD-O-3 Regional nodes positive (1988+)  \
300957             Malignant                               0   
58456              Malignant                               0   
429774             Malignant                               8   
226774             Malignant                               0   
194850             Malignant                               1   
...                      ...                             ...   
204005             Malignant                               1   
77221              Malignant                               0   
48914              Malignant   

In [67]:
print(test['Survival months'].value_counts())
print(train['Survival months'].value_counts())

-1    32221
 1    32181
Name: Survival months, dtype: int64
 1    128824
-1    128784
Name: Survival months, dtype: int64


# Preprocessing

### Refactor

so..... i dont think you need to hm. scale after you encode?????? or just dont do encoding and do one hot encoding for all of them

actually i think you can just try both of them and see which preprocessing turns out to be better 12/16/20

In [68]:
from sklearn.preprocessing import OrdinalEncoder
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import OneHotEncoder
from sklearn.preprocessing import MinMaxScaler
from sklearn.compose import ColumnTransformer
from sklearn.impute import SimpleImputer

In [69]:
class Preprocessing:
    from sklearn.preprocessing import OrdinalEncoder
    from sklearn.pipeline import Pipeline
    from sklearn.preprocessing import OneHotEncoder
    from sklearn.preprocessing import MinMaxScaler
    from sklearn.compose import ColumnTransformer
    from sklearn.impute import SimpleImputer
        
    def __init__(self, data):
        self.data = data
    
    def preprocess_all(self, scale_var, encode_var, hot_var):
        num = self.num_pipeline()
        encode = self.encode_scale_pipeline()
        one_hot = self.one_hot_pipeline()
        
        pipeline = ColumnTransformer([
        ('one_hot', one_hot, hot_var),
        ('encode', encode, encode_var),
        ('num', num, scale_var)
        ], remainder='passthrough')
        
        data_final = pipeline.fit_transform(self.data)
        return data_final
    
    def encode_scale_pipeline(self):
        encode_pipeline = Pipeline([
            ('encode', OrdinalEncoder()),
            ('scale', MinMaxScaler())
        ])
        return encode_pipeline
    
    def encode_pipeline(self):
        encode_pipeline = Pipeline([
            ('encode', OrdinalEncoder())
        ])
        return encode_pipeline    
    
    def num_pipeline(self):
        num_pipeline = Pipeline([
            ('imputer', SimpleImputer(strategy="median")),
            ('minmax', MinMaxScaler())
        ])
        return num_pipeline
    
    def one_hot_pipeline(self):
        one_hot_pipeline = Pipeline([
            ('cat_one_hot', OneHotEncoder())
        ])
        return one_hot_pipeline   

In [81]:
scale = ['CS tumor size (2004-2015)',  'Regional nodes positive (1988+)',  'Regional nodes examined (1988+)',  'Total number of in situ/malignant tumors for patient', 
         'Breast - Adjusted AJCC 6th T (1988-2015)', 'Breast - Adjusted AJCC 6th N (1988-2015)', 'Breast - Adjusted AJCC 6th M (1988-2015)']
#hot =  ['Laterality', 'Grade']
hot=[]
encode = ['Summary stage 2000 (1998+)', 'Grade',  'Race recode (W, B, AI, API)',  'Age recode with <1 year olds',  'Primary Site - labeled', 
           'Reason no cancer-directed surgery',  'Histologic Type ICD-O-3',  'Behavior code ICD-O-3',  'Laterality',
          'ER Status Recode Breast Cancer (1990+)', 'PR Status Recode Breast Cancer (1990+)', 'Breast - Adjusted AJCC 6th Stage (1988-2015)']

In [82]:
preprocess_1 = Preprocessing(train)
train_final = preprocess_1.preprocess_all(scale, encode, hot)

In [83]:
train_f = pd.DataFrame(train_final)

In [87]:
train_f

0         1     2       3      4    5         6    7     8   \
0       0.666667  0.000000  1.00  0.4375  0.500  1.0  0.756757  1.0  0.25   
1       0.666667  1.000000  1.00  0.8125  0.875  1.0  0.702703  1.0  1.00   
2       1.000000  0.333333  1.00  0.6250  1.000  1.0  0.756757  1.0  1.00   
3       0.666667  1.000000  1.00  0.6875  0.500  1.0  0.756757  1.0  0.25   
4       0.666667  0.333333  1.00  0.3750  0.500  1.0  0.756757  1.0  1.00   
...          ...       ...   ...     ...    ...  ...       ...  ...   ...   
257603  0.666667  1.000000  0.50  0.5625  0.500  1.0  0.810811  1.0  0.25   
257604  0.666667  0.333333  1.00  0.5625  0.375  1.0  0.756757  1.0  1.00   
257605  0.666667  0.333333  1.00  0.3750  0.500  1.0  0.756757  1.0  1.00   
257606  0.666667  0.333333  0.25  0.4375  0.375  1.0  0.324324  1.0  0.25   
257607  0.666667  1.000000  1.00  0.5625  0.875  1.0  0.810811  1.0  1.00   

              9         10        11        12        13        14        15  \
0       0.666667  0.666667  0.166667  0.007085  0.000000  0.000000  0.000000   
1       0.666667  0.666667  0.166667  0.009109  0.000000  0.011236  0.000000   
2       0.666667  0.666667  0.500000  0.029352  0.010101  0.022472  0.000000   
3       0.666667  0.666667  0.333333  0.031377  0.000000  0.000000  0.000000   
4       0.333333  0.333333  0.166667  0.017206  0.000000  0.000000  0.000000   
...          ...       ...       ...       ...       ...       ...       ...   
257603  0.666667  0.666667  0.166667  0.014170  0.000000  0.022472  0.000000   
257604  0.333333  0.333333  0.166667  0.011134  0.000000  0.022472  0.000000   
257605  0.666667  0.666667  0.166667  0.014170  0.000000  0.033708  0.000000   
257606  0.666667  0.333333  0.166667  0.014170  0.000000  0.033708  0.000000   
257607  0.666667  0.666667  0.166667  0.012146  0.000000  0.000000  0.083333   

              16   17   18   19  
0       0.083333  0.0  0.0  1.0  
1       0.083333  0.0  0.0  1.0  
2       0.333333  0.2  0.0 -1.0  
3       0.333333  0.0  0.0  1.0  
4       0.166667  0.0  0.0  1.0  
...          ...  ...  ...  ...  
257603  0.166667  0.0  0.0 -1.0  
257604  0.166667  0.0  0.0  1.0  
257605  0.166667  0.0  0.0  1.0  
257606  0.166667  0.0  0.0  1.0  
257607  0.166667  0.0  0.0  1.0  

[257608 rows x 20 columns]

In [85]:
preprocess_2 = Preprocessing(test)
test_final = preprocess_2.preprocess_all(scale, encode, hot)
test_f = pd.DataFrame(test_final)
test_f

0         1     2         3      4    5        6    7     8   \
0      0.666667  0.000000  1.00  0.642857  0.875  1.0  1.00000  1.0  1.00   
1      0.666667  0.333333  1.00  0.642857  1.000  1.0  0.71875  1.0  1.00   
2      1.000000  0.000000  0.25  0.357143  0.125  1.0  0.71875  1.0  1.00   
3      0.666667  0.000000  1.00  0.785714  0.875  1.0  0.71875  1.0  1.00   
4      1.000000  0.000000  1.00  0.428571  0.500  1.0  0.78125  1.0  1.00   
...         ...       ...   ...       ...    ...  ...      ...  ...   ...   
64397  1.000000  0.333333  0.25  0.428571  0.500  1.0  0.78125  1.0  0.25   
64398  0.666667  0.333333  1.00  0.714286  0.500  1.0  0.71875  1.0  0.25   
64399  0.666667  1.000000  1.00  0.928571  1.000  1.0  0.71875  1.0  1.00   
64400  0.666667  1.000000  1.00  0.428571  0.500  1.0  0.34375  1.0  1.00   
64401  0.666667  0.333333  1.00  0.642857  0.500  1.0  0.71875  1.0  1.00   

             9         10        11        12        13        14        15  \
0      1.000000  1.000000  0.333333  0.034413  0.000000  0.157303  0.000000   
1      0.666667  0.666667  0.166667  0.010121  0.000000  0.000000  0.000000   
2      0.333333  0.333333  0.666667  0.029352  0.080808  0.101124  0.142857   
3      0.666667  0.666667  0.166667  0.008097  0.000000  0.000000  0.000000   
4      0.666667  0.666667  0.500000  0.024291  0.010101  0.202247  0.285714   
...         ...       ...       ...       ...       ...       ...       ...   
64397  0.666667  0.666667  0.833333  0.044534  0.010101  0.011236  0.000000   
64398  0.666667  0.333333  0.333333  0.031377  0.000000  0.000000  0.000000   
64399  0.666667  0.666667  0.166667  0.007085  0.000000  0.000000  0.000000   
64400  0.666667  0.666667  0.166667  0.003036  0.000000  0.033708  0.000000   
64401  0.666667  0.666667  0.333333  0.028340  0.000000  0.056180  0.000000   

             16    17   18   19  
0      0.333333  0.00  0.0  1.0  
1      0.166667  0.00  0.0  1.0  
2      0.333333  0.65  0.0  1.0  
3      0.083333  0.00  0.0 -1.0  
4      0.333333  0.20  0.0  1.0  
...         ...   ...  ...  ...  
64397  0.777778  0.20  0.0  1.0  
64398  0.333333  0.00  0.0  1.0  
64399  0.083333  0.00  0.0 -1.0  
64400  0.027778  0.00  0.0  1.0  
64401  0.333333  0.00  0.0 -1.0  

[64402 rows x 20 columns]

data-2.txt has 448215 rows × 15 columns
data-3.txt has 439719 rows and 19 cols

data-4.txt has 257608 x 19 for train, 64402 x 19 for test

# WRITE TO TEXT FILE

In [88]:
train_f.to_csv('/cluster/2021mqian/snr research/recurrence_my_project/data-4-train.txt', sep=',', index=False, header=True)

In [89]:
test_f.to_csv('/cluster/2021mqian/snr research/recurrence_my_project/data-4-test.txt', sep=',', index=False, header=True)

# Training (old)

In [152]:
target = pd.DataFrame(target_1.to_numpy(), columns=['Survival months'])
target

Survival months
0                     1
1                     1
2                     1
3                     1
4                     1
...                 ...
448210                1
448211                1
448212                1
448213                1
448214                1

[448215 rows x 1 columns]

In [153]:
data_f

0         1     2       3      4    5         6    7     8   \
0       0.333333  0.666667  1.00  1.0000  1.000  1.0  0.611111  0.0  1.00   
1       0.333333  0.666667  1.00  0.6875  0.125  1.0  0.648148  0.0  0.25   
2       0.333333  0.333333  0.25  0.6250  0.500  1.0  0.648148  0.0  1.00   
3       0.333333  0.666667  1.00  0.8750  0.500  1.0  0.611111  0.0  0.25   
4       0.333333  0.000000  0.25  0.6875  0.250  1.0  0.648148  0.0  0.25   
...          ...       ...   ...     ...    ...  ...       ...  ...   ...   
448210  1.000000  0.000000  1.00  0.5625  0.500  1.0  0.648148  1.0  0.25   
448211  0.666667  0.000000  1.00  0.7500  0.875  1.0  0.611111  1.0  0.25   
448212  0.666667  0.000000  1.00  0.3125  0.250  1.0  0.611111  1.0  1.00   
448213  0.666667  1.000000  1.00  0.6250  0.250  1.0  0.648148  1.0  0.25   
448214  1.000000  0.333333  1.00  0.6250  0.875  1.0  0.611111  1.0  0.25   

              9         10        11        12        13        14  
0       0.333333  0.333333  0.017206  0.000000  0.000000  0.166667  
1       0.333333  0.333333  0.024291  0.000000  0.000000  0.083333  
2       0.333333  0.333333  0.029352  0.000000  0.000000  0.166667  
3       0.333333  0.333333  0.004049  0.000000  0.011236  0.166667  
4       0.666667  0.666667  0.032389  0.000000  0.011236  0.083333  
...          ...       ...       ...       ...       ...       ...  
448210  0.666667  0.666667  0.011134  0.101010  0.258427  0.000000  
448211  0.666667  0.666667  0.011134  0.000000  0.011236  0.000000  
448212  0.666667  0.666667  0.014170  0.000000  0.033708  0.000000  
448213  0.666667  0.333333  0.019231  0.000000  0.011236  0.000000  
448214  0.666667  0.666667  0.034413  0.121212  0.134831  0.000000  

[448215 rows x 15 columns]

In [154]:
data_all = pd.concat([data_f, target], axis=1)
data_all

0         1     2       3      4    5         6    7     8  \
0       0.333333  0.666667  1.00  1.0000  1.000  1.0  0.611111  0.0  1.00   
1       0.333333  0.666667  1.00  0.6875  0.125  1.0  0.648148  0.0  0.25   
2       0.333333  0.333333  0.25  0.6250  0.500  1.0  0.648148  0.0  1.00   
3       0.333333  0.666667  1.00  0.8750  0.500  1.0  0.611111  0.0  0.25   
4       0.333333  0.000000  0.25  0.6875  0.250  1.0  0.648148  0.0  0.25   
...          ...       ...   ...     ...    ...  ...       ...  ...   ...   
448210  1.000000  0.000000  1.00  0.5625  0.500  1.0  0.648148  1.0  0.25   
448211  0.666667  0.000000  1.00  0.7500  0.875  1.0  0.611111  1.0  0.25   
448212  0.666667  0.000000  1.00  0.3125  0.250  1.0  0.611111  1.0  1.00   
448213  0.666667  1.000000  1.00  0.6250  0.250  1.0  0.648148  1.0  0.25   
448214  1.000000  0.333333  1.00  0.6250  0.875  1.0  0.611111  1.0  0.25   

               9        10        11        12        13        14  \
0       0.333333  0.333333  0.017206  0.000000  0.000000  0.166667   
1       0.333333  0.333333  0.024291  0.000000  0.000000  0.083333   
2       0.333333  0.333333  0.029352  0.000000  0.000000  0.166667   
3       0.333333  0.333333  0.004049  0.000000  0.011236  0.166667   
4       0.666667  0.666667  0.032389  0.000000  0.011236  0.083333   
...          ...       ...       ...       ...       ...       ...   
448210  0.666667  0.666667  0.011134  0.101010  0.258427  0.000000   
448211  0.666667  0.666667  0.011134  0.000000  0.011236  0.000000   
448212  0.666667  0.666667  0.014170  0.000000  0.033708  0.000000   
448213  0.666667  0.333333  0.019231  0.000000  0.011236  0.000000   
448214  0.666667  0.666667  0.034413  0.121212  0.134831  0.000000   

        Survival months  
0                     1  
1                     1  
2                     1  
3                     1  
4                     1  
...                 ...  
448210                1  
448211                1  
448212                1  
448213                1  
448214                1  

[448215 rows x 16 columns]

Resampling --> having the same number of survived and did not survive

In [155]:
data_all['Survival months'].value_counts()

 1    281919
-1    166296
Name: Survival months, dtype: int64

In [156]:
low = data_all[['Survival months']].value_counts().min()
low

166296

Choosing the number of random samples that have (data-2 166296)  (data-3 161005) (data-1 171866) for the survived (1) cases

In [157]:
samples = data_all[data_all['Survival months'] == 1].index.values
index_range = range(samples.shape[0])
indexes = np.random.choice(index_range, size=low, replace=False)
subsample =  samples[indexes].tolist()
d_1 = data_all.loc[subsample, :]

In [158]:
d_1

0         1    2       3      4    5         6    7     8  \
60822   0.666667  0.333333  1.0  0.3750  0.625  1.0  0.611111  1.0  0.25   
152243  1.000000  0.333333  1.0  0.8750  0.125  1.0  0.611111  1.0  0.25   
71082   0.666667  0.000000  1.0  0.7500  1.000  1.0  0.648148  1.0  1.00   
309807  0.666667  1.000000  1.0  0.7500  0.875  1.0  0.611111  1.0  0.25   
283555  0.666667  1.000000  1.0  0.8750  0.500  1.0  0.611111  1.0  1.00   
...          ...       ...  ...     ...    ...  ...       ...  ...   ...   
425004  0.666667  0.000000  1.0  0.6875  0.500  1.0  0.611111  1.0  0.25   
311752  0.666667  1.000000  1.0  0.8750  0.875  1.0  0.648148  1.0  0.25   
415387  0.666667  0.333333  1.0  0.7500  0.250  1.0  0.611111  1.0  1.00   
254028  0.666667  0.000000  1.0  0.6875  0.875  1.0  0.648148  1.0  0.25   
245883  1.000000  0.333333  1.0  0.6875  0.625  1.0  0.611111  1.0  0.25   

               9        10        11        12        13        14  \
60822   0.333333  0.333333  0.026316  0.000000  0.011236  0.000000   
152243  0.666667  0.333333  0.030364  0.010101  0.146067  0.000000   
71082   0.666667  0.666667  0.011134  0.000000  0.000000  0.083333   
309807  0.666667  0.666667  0.013158  0.000000  0.011236  0.000000   
283555  0.666667  0.666667  0.006073  0.000000  0.044944  0.000000   
...          ...       ...       ...       ...       ...       ...   
425004  0.666667  0.666667  0.006073  0.000000  0.022472  0.000000   
311752  0.333333  0.333333  0.008097  0.000000  0.044944  0.000000   
415387  0.666667  0.666667  0.009109  0.000000  0.011236  0.000000   
254028  0.666667  0.666667  0.055668  0.000000  0.022472  0.000000   
245883  0.666667  0.333333  0.024291  0.010101  0.202247  0.000000   

        Survival months  
60822                 1  
152243                1  
71082                 1  
309807                1  
283555                1  
...                 ...  
425004                1  
311752                1  
415387                1  
254028                1  
245883                1  

[166296 rows x 16 columns]

In [159]:
samples_0 = data_all[data_all['Survival months'] == -1].index.values
d_0 = data_all.loc[samples_0, :]

In [160]:
d_0

0         1     2       3      4    5         6    7     8  \
6       0.333333  0.000000  0.25  0.6875  0.500  1.0  0.648148  0.0  0.25   
17      0.333333  0.333333  0.50  0.5625  0.500  1.0  0.648148  0.0  1.00   
18      0.333333  0.333333  0.25  0.6250  0.875  1.0  0.611111  0.0  0.25   
24      0.333333  0.000000  0.50  0.6250  1.000  1.0  0.648148  0.0  1.00   
33      0.333333  0.000000  1.00  0.4375  0.250  1.0  0.611111  0.0  1.00   
...          ...       ...   ...     ...    ...  ...       ...  ...   ...   
448173  0.666667  0.333333  1.00  0.9375  0.500  1.0  0.611111  1.0  0.25   
448178  1.000000  0.000000  0.25  0.5625  0.125  1.0  0.611111  1.0  1.00   
448183  0.000000  0.000000  1.00  0.5000  0.500  1.0  0.611111  1.0  0.25   
448190  0.666667  0.000000  0.25  0.8125  0.875  1.0  0.611111  1.0  0.25   
448200  0.666667  0.000000  1.00  0.5000  1.000  1.0  0.611111  1.0  0.25   

               9        10        11        12        13        14  \
6       0.666667  0.666667  0.007085  0.000000  0.000000  0.083333   
17      0.666667  0.666667  0.024291  0.000000  0.011236  0.166667   
18      0.666667  0.666667  0.019231  0.000000  0.000000  0.083333   
24      0.666667  0.666667  0.049595  0.000000  0.011236  0.083333   
33      0.666667  0.666667  0.017206  0.000000  0.011236  0.083333   
...          ...       ...       ...       ...       ...       ...   
448173  0.666667  0.666667  0.016194  0.000000  0.056180  0.000000   
448178  0.666667  0.666667  0.039474  0.050505  0.056180  0.000000   
448183  0.666667  0.666667  0.009109  0.000000  0.123596  0.000000   
448190  0.666667  0.666667  0.007085  0.000000  0.011236  0.000000   
448200  0.666667  0.666667  0.014170  0.000000  0.044944  0.000000   

        Survival months  
6                    -1  
17                   -1  
18                   -1  
24                   -1  
33                   -1  
...                 ...  
448173               -1  
448178               -1  
448183               -1  
448190               -1  
448200               -1  

[166296 rows x 16 columns]

In [161]:
d_final = pd.concat([d_1, d_0])

In [162]:
d_final[['Survival months']].value_counts()

Survival months
 1                 166296
-1                 166296
dtype: int64

In [163]:
d_final

0         1     2       3      4    5         6    7     8  \
60822   0.666667  0.333333  1.00  0.3750  0.625  1.0  0.611111  1.0  0.25   
152243  1.000000  0.333333  1.00  0.8750  0.125  1.0  0.611111  1.0  0.25   
71082   0.666667  0.000000  1.00  0.7500  1.000  1.0  0.648148  1.0  1.00   
309807  0.666667  1.000000  1.00  0.7500  0.875  1.0  0.611111  1.0  0.25   
283555  0.666667  1.000000  1.00  0.8750  0.500  1.0  0.611111  1.0  1.00   
...          ...       ...   ...     ...    ...  ...       ...  ...   ...   
448173  0.666667  0.333333  1.00  0.9375  0.500  1.0  0.611111  1.0  0.25   
448178  1.000000  0.000000  0.25  0.5625  0.125  1.0  0.611111  1.0  1.00   
448183  0.000000  0.000000  1.00  0.5000  0.500  1.0  0.611111  1.0  0.25   
448190  0.666667  0.000000  0.25  0.8125  0.875  1.0  0.611111  1.0  0.25   
448200  0.666667  0.000000  1.00  0.5000  1.000  1.0  0.611111  1.0  0.25   

               9        10        11        12        13        14  \
60822   0.333333  0.333333  0.026316  0.000000  0.011236  0.000000   
152243  0.666667  0.333333  0.030364  0.010101  0.146067  0.000000   
71082   0.666667  0.666667  0.011134  0.000000  0.000000  0.083333   
309807  0.666667  0.666667  0.013158  0.000000  0.011236  0.000000   
283555  0.666667  0.666667  0.006073  0.000000  0.044944  0.000000   
...          ...       ...       ...       ...       ...       ...   
448173  0.666667  0.666667  0.016194  0.000000  0.056180  0.000000   
448178  0.666667  0.666667  0.039474  0.050505  0.056180  0.000000   
448183  0.666667  0.666667  0.009109  0.000000  0.123596  0.000000   
448190  0.666667  0.666667  0.007085  0.000000  0.011236  0.000000   
448200  0.666667  0.666667  0.014170  0.000000  0.044944  0.000000   

        Survival months  
60822                 1  
152243                1  
71082                 1  
309807                1  
283555                1  
...                 ...  
448173               -1  
448178               -1  
448183               -1  
448190               -1  
448200               -1  

[332592 rows x 16 columns]

In [164]:
d_shuffle = d_final.sample(frac=1)

In [165]:
d_shuffle

0         1     2       3      4    5         6    7     8  \
335422  0.666667  0.333333  0.25  0.4375  0.500  1.0  0.611111  1.0  1.00   
421787  1.000000  0.333333  1.00  0.6875  1.000  1.0  0.611111  1.0  1.00   
162238  0.666667  0.000000  1.00  0.6875  0.250  1.0  0.648148  1.0  0.25   
79290   0.666667  0.000000  1.00  0.6250  1.000  1.0  0.611111  1.0  0.25   
258253  1.000000  0.333333  1.00  0.6875  0.125  1.0  0.648148  1.0  0.25   
...          ...       ...   ...     ...    ...  ...       ...  ...   ...   
420594  0.666667  0.000000  1.00  0.5000  0.375  1.0  0.611111  1.0  0.25   
424927  0.666667  0.333333  1.00  0.6250  0.875  1.0  0.611111  1.0  0.25   
429924  1.000000  0.000000  1.00  0.6875  0.500  1.0  0.611111  1.0  0.25   
193406  1.000000  0.000000  1.00  0.7500  0.500  1.0  0.648148  1.0  0.25   
69038   1.000000  1.000000  1.00  0.9375  0.500  1.0  0.611111  1.0  0.25   

               9        10        11        12        13        14  \
335422  0.333333  0.333333  0.011134  0.000000  0.022472  0.083333   
421787  0.666667  0.666667  0.033401  0.010101  0.056180  0.000000   
162238  0.666667  0.666667  0.018219  0.000000  0.022472  0.000000   
79290   0.666667  0.666667  0.008097  0.000000  0.000000  0.000000   
258253  0.333333  0.333333  0.011134  0.010101  0.382022  0.000000   
...          ...       ...       ...       ...       ...       ...   
420594  0.666667  0.666667  0.010121  0.000000  0.022472  0.000000   
424927  0.333333  0.333333  0.019231  0.000000  0.089888  0.000000   
429924  0.666667  0.666667  0.012146  0.010101  0.123596  0.000000   
193406  0.666667  0.666667  0.024291  0.020202  0.280899  0.000000   
69038   0.666667  0.666667  0.008097  0.040404  0.067416  0.000000   

        Survival months  
335422                1  
421787                1  
162238                1  
79290                -1  
258253                1  
...                 ...  
420594                1  
424927                1  
429924                1  
193406                1  
69038                -1  

[332592 rows x 16 columns]